apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,614 @@
+			(
+				user_id,
+				last_failed_at
+			)
+			values
+			(
+				event->'user_id',
+				now()
+			)
+			on conflict
+        do update
+          set last_failed_at = now();
+
+		-- finally let Supabase Auth do the default behavior for a failed attempt
+		return jsonb_build_object('decision', 'continue');
+	end;
+	$$;
+
+
+-- Assign appropriate permissions
+grant execute
+	on function public.hook_password_verification_attempt
+	to supabase_auth_admin;
+
+grant all
+	on table public.password_failed_verification_attempts
+	to supabase_auth_admin;
+
+revoke execute
+	on function public.hook_password_verification_attempt
+	from authenticated, anon;
+
+revoke all
+	on table public.password_failed_verification_attempts
+	from authenticated, anon;
+```
+</TabPanel>
+</Tabs>
+
+## Custom Access Token Hook
+
+The custom access token hook runs before a token is issued and allows you to add additional claims based on the authentication method used or add additional metadata to select users.
+
+Claims returned must conform to our specification. Supabase Auth will check for these claims after the hook is run and return an error if they are not present.
+
+Here's a listing of all fields currently available on an access token:
+
+| Field           | Is Required Claim |
+|-----------------|-------------------|
+| `aud`           | Yes               |
+| `exp`           | Yes               |
+| `jti`           | No                |
+| `iat`           | Yes               |
+| `iss`           | No                |
+| `nbf`           | No                |
+| `sub`           | Yes               |
+| `email`         | Yes               |
+| `phone`         | Yes               |
+| `app_metadata`  | No                |
+| `user_metadata` | No                |
+| `role`          | Yes               |
+| `aal`           | Yes               |
+| `amr`           | No                |
+| `session_id`    | No                |
+
+
+### Inputs
+
+- `user_id`Unique identifier for the user attempting to sign in. Correlate this to the `auth.users` table.
+- `claims` Claims which are attached to the access token.
+- `authentication_method` the authentication method used to request the access token. Possible values include:
+`oauth`, `password`, `otp`, `totp`, `recovery`, `invite`, `sso/saml`, `magiclink`, `email/signup`, `email_change`, `token_refresh`.
+
+
+
+### Outputs
+
+Return these only if your hook processed the input without errors.
+
+- `claims` A json containing the updated claims after the hook has been run.
+
+### Example: Add Custom Claims to Token
+
+<Tabs
+  scrollable
+  size="small"
+  type="underlined"
+  defaultActiveId="js"
+  queryGroup="language"
+>
+<TabPanel id="add-admin-role" label="Add Admin Role">
+
+Your company would like to allow users with a `@supabase.com` email account to perform restricted actions. To support this, users who are registered as admins with the company should be accorded the `admin` role.
```

Comment:
from what I can see the function doesn't currently check if the users are admins right? it just automatically makes everybody with a `@supabase.com` email an admin?